### PR TITLE
fix: panic in worker when closing at top level

### PR DESCRIPTION
**Pull request by bartlomieju**
Fixes panic occurring in worker when `self.close()` is called
at the top level, ie. worker shuts down while
module evaluation promise hasn't yet resolved.

Closes https://github.com/denoland/deno/issues/7620
Closes https://github.com/denoland/deno/issues/5334

Supersedes https://github.com/denoland/deno/pull/8395

### CODE REVIEW

File: cli/tests/no_mem_cache.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
+const fixtureFile = await Deno.makeTempFile();
+let prefix = "file://";
+if (Deno.build.os == "windows") {
+  prefix += "/";
+}
```
Comment (ry) [inline cli/tests/no_mem_cache.js:5]: Aside: I wish had something like `Deno.makeTempFile().toUrl()`

File: core/runtime.rs [hunk 2]
```diff
@@ -901,9 +901,13 @@ impl JsRuntime {
...
-      if let Poll::Ready(result) = receiver.poll_next_unpin(cx) {
-        debug!("received module evaluate");
-        return Poll::Ready(result.unwrap());
+      if let Poll::Ready(maybe_result) = receiver.poll_next_unpin(cx) {
+        debug!("received module evaluate {:#?}", maybe_result);
+        // If `None` is returned it means that runtime was destroyed before
+        // evaluation was complete. This can happen in Web Worker when `self.close()`
+        // is called at top level.
+        let result = maybe_result.unwrap_or(Ok(()));
+        return Poll::Ready(result);
```
Comment (ry) [inline core/runtime.rs:910]: It would be really great to have a standalone test in core for this...

File: cli/tests/no_mem_cache.js (hunk 1 above)
Comment (bartlomieju) [reply cli/tests/no_mem_cache.js:5]: @ry IMO this should be handled by `URL`. This bit is very problematic in other contexts, especially in dynamic imports: https://github.com/denoland/deno/issues/4641

Unfortunately JS `URL` can't resolve it properly just like `Url` Rust crate 😢

**Review by ry — APPROVED** [review]
Discussed offline - too difficult to get test in core.

LGTM

**Pull request merged**